src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,292 @@
+Prf: r being a root implies that (2*r.L)//r^2 is a subset of ZZ
+     r being primitive implies that 2*i*ZZ is a subset of 2*r.L,
+     which is a subset of (r^2)*ZZ, what implies that r^2 divides 2*i
+"""
+function _all_root_lengths(Q::ZZMatrix)
+  l = nrows(Q)
+  S = snf(Q)
+  return (-1) * divisors(2 * S[l, l])
+end
+
+@doc raw"""
+    _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem}) -> Vector{ZZRingElem}
+
+Check whether the given lengths are possible root lengths.
+Unnecessary roots are sorted out and reported if this is wanted.
+Return only the possible root lengths.
+"""
+function _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem})
+  l = nrows(Q)
+  S = snf(Q)
+  possible_root_lengths = divisors(2 * S[l, l])
+  if isempty(root_lengths)
+    return possible_root_lengths
+  end
+  result = ZZRingElem[]
+  for r in root_lengths
+    if abs(r) in possible_root_lengths
+      push!(result, r)
+    else
+      @vprintln :Vinberg 1 "$r is not a possible root length"
+    end
+  end
+  @req !isempty(result) "No possible root lengths found"
+  return result
+end
+
+@doc raw"""
+    _distance_indices(upper_bound::ZZRingElem, root_lengths::Vector{ZZRingElem}) -> Vector{Tuple{Int, ZZRingElem}}
+
+Returns all possible tuples (n, l) with n a natural number and l contained in root_lengths,
+sorted by increase of the value n^2//l, stopping by upper_bound.
+"""
+function _distance_indices(upper_bound::ZZRingElem, root_lengths::Vector{ZZRingElem})
+
+  result = Tuple{Int,ZZRingElem}[]
+
+  for l in root_lengths
+    x0 = sqrt(Float64(abs(l * upper_bound))) # consider n^2//l < upper_bound => n < sqrt(l*upper_bound)
+    x = Int(floor(x0))
+    for n in 1:x
+      push!(result, (n, l))
+    end
+  end
+
+  sort!(result; by=x -> (x[1]^2) // abs(x[2]))
+  return result
+end
+
+@doc raw"""
+    _is_primitive(v)
+
+Check if v is primitive, which is equivalent to checking whether the gcd of the entries of v is equal to 1.
+"""
+function _is_primitive(v)
+  return isone(reduce(gcd, v))
+end
+
+@doc raw"""
+    _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+
+Check if reflection by v preserves the lattice, i.e. check if for v a row vector
+(2*v*Q)//v^2 is an integer matrix. 
+"""
+function _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+  A = Q * transpose(v)
+  b = (v*A)[1, 1]
+  A = 2 * A
+
+  return all(iszero(mod(x, b)) for x in A)
+end
+
+@doc raw"""
+    _check_coorientation(Q::ZZMatrix, roots::Vector{ZZMatrix}, v::ZZMatrix, v0::ZZMatrix)
+
+First check whether v has non-obtuse angles with all roots r already found by checking that v.r >= 0.
+Then also check the coorientation of v by checking that v.v0 >= 0.
+"""
+function _check_coorientation(Q::ZZMatrix, roots::Vector{ZZMatrix}, v::ZZMatrix, v0::ZZMatrix)
+  Qv = Q * transpose(v)
+  for r in roots
+    if (r*Qv)[1, 1] < 0
+      return false
+    end
+  end
+
+  if (v*Q*transpose(v0))[1, 1] < 0
+    return false
+  end
+  return true
```

Comment:
could be shortened to 
```suggestion
  return (v*Q*transpose(v0))[1, 1] > 0
```